File: android/app/src/main/java/com/tonhub/wallet/push/PushNotificationManager.java
package com.tonhub.wallet.push;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Manager for handling push notifications
 */
public class PushNotificationManager {

    private static final String TAG = "PushNotificationMgr";
    private static Bundle lastPushData = null;
    private static boolean waitingToEmitEvent = false;
    private static boolean isFirstLaunch = true;
    private static PushNotificationManager instance;

    private final ReactApplication mReactApplication;

    private PushNotificationManager(ReactApplication reactApplication) {
        mReactApplication = reactApplication;
    }

    public static synchronized PushNotificationManager getInstance(ReactApplication reactApplication) {
        if (instance == null) {
            instance = new PushNotificationManager(reactApplication);
        }
        return instance;
    }

    /**
     * Processes the push notification intent
     */
    public void processNotificationIntent(Intent intent) {
        if (intent == null)
            return;

        Bundle extras = intent.getExtras();
        if (extras != null) {
            lastPushData = (Bundle) extras.clone();
        }
    }

    /**
     * Called on activity onStart, tracks the first launch of the app
     */
    public void onStart() {
        if (!isFirstLaunch) {
            emitNotificationDataIfNeeded();
        }
        isFirstLaunch = false;
    }

    /**
     * Called on activity onResume
     */
    public void onResume() {
        if (!isFirstLaunch) {
            emitNotificationDataIfNeeded();
        }
    }

    /**
     * Processes new intents that may contain push notification data
     */
    public void onNewIntent(Intent intent) {
        processNotificationIntent(intent);
        emitNotificationDataIfNeeded();
    }

    /**
     * Sends data to JavaScript if needed
     */
    private void emitNotificationDataIfNeeded() {
        if (lastPushData == null) {
            return;
        }

        try {
            final ReactInstanceManager reactInstanceManager = mReactApplication.getReactNativeHost()
                    .getReactInstanceManager();
            ReactContext reactContext = reactInstanceManager.getCurrentReactContext();

            if (reactContext != null) {
                sendDataToJS(reactContext, lastPushData);
                lastPushData = null;
                waitingToEmitEvent = false;
            } else {
                waitingToEmitEvent = true;
                reactInstanceManager
                        .addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
                            @Override
                            public void onReactContextInitialized(ReactContext context) {
                                sendDataToJS(context, lastPushData);
                                lastPushData = null;
                                waitingToEmitEvent = false;
                                reactInstanceManager.removeReactInstanceEventListener(this);
                            }
                        });
            }
        } catch (Exception e) {
            if (!waitingToEmitEvent) {
                waitingToEmitEvent = true;
                new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        waitingToEmitEvent = false;
                        emitNotificationDataIfNeeded();
                    }
                }, 1000);
            }
        }
    }

    /**
     * Sends data to JavaScript
     */
    private void sendDataToJS(ReactContext reactContext, Bundle data) {
        try {
            WritableMap params = Arguments.createMap();
            String url = extractUrlFromBundle(data);
            if (url != null) {
                params.putString("url", url);
            } else {
                bundleToWritableMap(data, params);
            }
            reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit("pushNotificationOpened", params);
        } catch (Exception e) {
            Log.e(TAG, "Error sending  notification data to JS: ", e);
        }
    }

    /**
     * Converts Bundle to WritableMap for passing to JS
     */
    private void bundleToWritableMap(Bundle bundle, WritableMap map) {
        for (String key : bundle.keySet()) {
            Object value = bundle.get(key);
            if (value == null) {
                map.putNull(key);
            } else if (value instanceof String) {
                map.putString(key, (String) value);
            } else if (value instanceof Integer) {
                map.putInt(key, (Integer) value);
            } else if (value instanceof Boolean) {
                map.putBoolean(key, (Boolean) value);
            } else if (value instanceof Double) {
                map.putDouble(key, (Double) value);
            } else {
                map.putString(key, value.toString());
            }
        }
    }

    /**
     * Extracts URL from various fields in Bundle
     */
    private String extractUrlFromBundle(Bundle data) {
        // Direct url field
        if (data.containsKey("url")) {
            return data.getString("url");
        }

        // Check data field
        String url = extractUrlFromJsonField(data, "data");
        if (url != null) {
            return url;
        }

        // Check body field
        url = extractUrlFromJsonField(data, "body");
        if (url != null) {
            return url;
        }

        // Check notification field
        if (data.containsKey("notification")) {
            try {
                String notifStr = data.getString("notification");
                JSONObject notifJson = new JSONObject(notifStr);
                if (notifJson.has("data") && notifJson.getJSONObject("data").has("url")) {
                    return notifJson.getJSONObject("data").getString("url");
                } else if (notifJson.has("url")) {
                    return notifJson.getString("url");
                }
            } catch (Exception e) {
                // Ignore errors
            }
        }

        return null;
    }

    /**
     * Helper method to extract URL from a JSON string field in Bundle
     */
    private String extractUrlFromJsonField(Bundle data, String fieldName) {
        if (data.containsKey(fieldName)) {
            try {
                String fieldValue = data.getString(fieldName);
                if (fieldValue != null) {
                    try {
                        JSONObject jsonObj = new JSONObject(fieldValue);
                        if (jsonObj.has("url")) {
                            return jsonObj.getString("url");
                        }
                    } catch (JSONException e) {
                        return fieldValue;
                    }
                }
            } catch (Exception e) {
                // Ignore errors
            }
        }
        return null;
    }
}

File: android/app/src/main/java/com/tonhub/wallet/push/PushNotificationManager.kt
package com.tonhub.wallet.push

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

/**
 * Manager for handling push notifications
 */
class PushNotificationManager private constructor(
    private val mReactApplication: ReactApplication
) {
    companion object {
        private const val TAG = "PushNotificationMgr"
        private var lastPushData: Bundle? = null
        private var waitingToEmitEvent = false
        private var isFirstLaunch = true
        private var instance: PushNotificationManager? = null

        @Synchronized
        fun getInstance(reactApplication: ReactApplication): PushNotificationManager {
            if (instance == null) {
                instance = PushNotificationManager(reactApplication)
            }
            return instance!!
        }
    }

    /**
     * Processes the push notification intent
     */
    fun processNotificationIntent(intent: Intent?) {
        if (intent == null) return

        val extras = intent.extras
        if (extras != null) {
            lastPushData = Bundle(extras)
        }
    }

    /**
     * Called on activity onStart, tracks the first launch of the app
     */
    fun onStart() {
        if (!isFirstLaunch) {
            emitNotificationDataIfNeeded()
        }
        isFirstLaunch = false
    }

    /**
     * Called on activity onResume
     */
    fun onResume() {
        if (!isFirstLaunch) {
            emitNotificationDataIfNeeded()
        }
    }

    /**
     * Processes new intents that may contain push notification data
     */
    fun onNewIntent(intent: Intent?) {
        processNotificationIntent(intent)
        emitNotificationDataIfNeeded()
    }

    /**
     * Sends data to JavaScript if needed
     */
    private fun emitNotificationDataIfNeeded() {
        val data = lastPushData ?: return

        try {
            val reactInstanceManager = mReactApplication.reactNativeHost.reactInstanceManager
            val reactContext = reactInstanceManager.currentReactContext

            if (reactContext != null) {
                sendDataToJS(reactContext, data)
                lastPushData = null
                waitingToEmitEvent = false
            } else {
                waitingToEmitEvent = true
                reactInstanceManager.addReactInstanceEventListener(object :
                    ReactInstanceManager.ReactInstanceEventListener {
                    override fun onReactContextInitialized(context: ReactContext) {
                        lastPushData?.let { sendDataToJS(context, it) }
                        lastPushData = null
                        waitingToEmitEvent = false
                        reactInstanceManager.removeReactInstanceEventListener(this)
                    }
                })
            }
        } catch (e: Exception) {
            if (!waitingToEmitEvent) {
                waitingToEmitEvent = true
                Handler(Looper.getMainLooper()).postDelayed({
                    waitingToEmitEvent = false
                    emitNotificationDataIfNeeded()
                }, 1000)
            }
        }
    }

    /**
     * Sends data to JavaScript
     */
    private fun sendDataToJS(reactContext: ReactContext, data: Bundle) {
        try {
            val params = Arguments.createMap()
            bundleToWritableMap(data, params)
            reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("pushNotificationOpened", params)
        } catch (e: Exception) {
            Log.e(TAG, "Error sending notification data to JS: ", e)
        }
    }

    /**
     * Converts Bundle to WritableMap for passing to JS
     */
    private fun bundleToWritableMap(bundle: Bundle, map: WritableMap) {
        for (key in bundle.keySet()) {
            when (val value = bundle.get(key)) {
                null -> map.putNull(key)
                is String -> map.putString(key, value)
                is Int -> map.putInt(key, value)
                is Boolean -> map.putBoolean(key, value)
                is Double -> map.putDouble(key, value)
                else -> map.putString(key, value.toString())
            }
        }
    }
}

